File: Harness.java
import java.util.*;
import java.util.logging.Level;
import java.util.logging.Logger;
import org.json.simple.JSONArray;
import org.json.simple.JSONObject;
import org.json.simple.parser.JSONParser;
import main.MainClass$;
import java.io.StringWriter;
import java.io.PrintWriter;
import scala.Tuple2;
import java.io.*;
import io.circe.Json;
import java.util.jar.Manifest;

import com.fasterxml.jackson.annotation.JsonProperty;
import com.fasterxml.jackson.core.JsonProcessingException;
import com.fasterxml.jackson.core.type.TypeReference;
import com.fasterxml.jackson.databind.DeserializationFeature;
import com.fasterxml.jackson.databind.JsonNode;
import com.fasterxml.jackson.databind.ObjectMapper;

/**
* The Harness program implements an application that
* simply take input from Bowtie application using standard
* input and output to the standard output.
*
* @author  Sajal Jain
* @version 1.0
* @since   2023-01-10 
*/
class Harness{
	
	public static boolean started = false;
	
	private static final Logger LOGGER = Logger.getLogger( Harness.class.getName() );
	
	private final ObjectMapper objectMapper = new ObjectMapper().configure(
      DeserializationFeature.FAIL_ON_UNKNOWN_PROPERTIES, false);
	
	private static final String NOT_IMPLEMENTED = "This case is not yet implemented.";
	private static final Map<String, String> UNSUPPORTED_CASES = Map.ofEntries(
		Map.entry("escaped pointer ref", NOT_IMPLEMENTED), Map.entry("empty tokens in $ref json-pointer", NOT_IMPLEMENTED),
		Map.entry("schema that uses custom metaschema with with no validation vocabulary", NOT_IMPLEMENTED),
		Map.entry("small multiple of large integer", NOT_IMPLEMENTED), Map.entry("$ref to $ref finds detached $anchor", NOT_IMPLEMENTED),
		Map.entry("$ref to $dynamicRef finds detached $dynamicAnchor", NOT_IMPLEMENTED));
    private static final Map<String, SpecificSkip> UNSUPPORTED_CASES_SPECIFIC = Map.ofEntries(
        Map.entry("minLength validation", new SpecificSkip("one supplementary Unicode code point is not long enough", NOT_IMPLEMENTED)), 
        Map.entry("maxLength validation", new SpecificSkip("two supplementary Unicode code points is long enough", NOT_IMPLEMENTED)));
	
	public static void main(String[] args) {
		Scanner input = new Scanner(System.in);
		while (true) {
			String line = input.nextLine();
			String output = new Harness().operate(line);
			System.out.println(output);
		}
	}
	
	public String operate(String line){
		String error = "";
		try{
			JsonNode node = objectMapper.readTree(line);
			String cmd = node.get("cmd").asText();
			switch(cmd) {
			  case "start":
				StartRequest startRequest = objectMapper.treeToValue(node, StartRequest.class);
				long version = startRequest.version();
				if(version == 1){
					InputStream is = getClass().getResourceAsStream("META-INF/MANIFEST.MF");
					var attributes = new Manifest(is).getMainAttributes();
					started = true;
					JSONObject message = new JSONObject();
					message.put("ready", true);
					message.put("version", 1);
					JSONObject implementation = new JSONObject();
					implementation.put("language", "scala");
					implementation.put("name", attributes.getValue("Implementation-Name"));
					implementation.put("version", attributes.getValue("Implementation-Version"));
					implementation.put("homepage", "https://gitlab.lip6.fr/jsonschema/modernjsonschemavalidator");
					implementation.put("issues", "https://gitlab.lip6.fr/jsonschema/modernjsonschemavalidator/issues");
					JSONArray dialects = new JSONArray();
					dialects.add("https://json-schema.org/draft/2020-12/schema");
					implementation.put("dialects", dialects);
					message.put("implementation", implementation);
					return message.toJSONString();
				}
				break;
			  case "dialect":
				if(started != true){
					throw new RuntimeException("Bowtie hasn't started!");
				}
				DialectRequest dialectRequest = objectMapper.treeToValue(node, DialectRequest.class);
				return "{ \"ok\" : false }";
			  case "run":
				if(started != true){
					throw new RuntimeException("Bowtie hasn't started!");
				}
				RunRequest runRequest = objectMapper.treeToValue(node, RunRequest.class);
				
				try{
                    String caseDescription = runRequest.testCase().description();
					if(UNSUPPORTED_CASES.containsKey(caseDescription)){
						return skipMsg(UNSUPPORTED_CASES.get(caseDescription), runRequest.seq().asLong());
					} 
					
					JSONArray resultArray = new JSONArray();
					
					for (Test test : runRequest.testCase().tests()) {
                        if (UNSUPPORTED_CASES_SPECIFIC.containsKey(caseDescription)) {
                            SpecificSkip skip = UNSUPPORTED_CASES_SPECIFIC.get(caseDescription);
                            if (skip.description().equals(test.description())) {
                                resultArray.add(skipMsg(UNSUPPORTED_CASES_SPECIFIC.get(caseDescription).message()));
                                continue;
                            }
                        }
						String instance = test.instance().toString();
						MainClass$ m = MainClass$.MODULE$;
						boolean results = m.validateInstance(runRequest.testCase().schema().toString(), instance);
						
						JSONObject result = new JSONObject();
						result.put("valid", (boolean) results);
						resultArray.add(result);
					}
					
					JSONObject out = new JSONObject();
					out.put("seq", runRequest.seq().asLong());
					out.put("results", resultArray);
					return out.toJSONString();
				}
				catch(Exception e){
					String msg = getDetailedMessage(e, runRequest.testCase().schema().toString());
					error = errorMsg(msg, runRequest.seq().asLong());
					return error;
				}
			  case "stop":
				if(started != true){
					throw new RuntimeException("Bowtie hasn't started!");
				}
				System.exit(0);
			}
		}catch(Exception e){
			error = errorMsg(e.getMessage(), -1);
			return error;
		}
		error = errorMsg("Send correct command", -1);
		return error;
	}
	
	public static String errorMsg(String message, long seq){
		JSONObject traceBack = new JSONObject(); 
		traceBack.put("traceBack", message);
		JSONObject error = new JSONObject(); 
		error.put("errored", true);
		error.put("seq", seq);
		error.put("context",traceBack);
		return error.toJSONString();
	}
	
	public static String skipMsg(String message, long seq){
		JSONObject error = new JSONObject();
		error.put("skipped", true);
		error.put("seq", seq);
		error.put("message",message);
		return error.toJSONString();
	}
	
    public static JSONObject skipMsg(String message){
		JSONObject error = new JSONObject();
		error.put("skipped", true);
		error.put("message",message);
		return error;
	}
	
    
	public static String getDetailedMessage(Exception e, String schema){
		StringWriter sw = new StringWriter();
		e.printStackTrace(new PrintWriter(sw));
		return sw.toString() + " " + schema;
	}
}

record StartRequest(int version) {}

record DialectRequest(String dialect) {}

record RunRequest(JsonNode seq, @JsonProperty("case") TestCase testCase) {}

record TestCase(String description, String comment, JsonNode schema, JsonNode registry, List<Test> tests) {}

record Test(String description, String comment, JsonNode instance, boolean valid) {}

record SpecificSkip(String description, String message) {}

File: Harness.scala
// TODO: Prints
import java.io._
import io.circe.generic.auto._
import io.circe.generic.semiauto._
import io.circe.parser.parse
import io.circe.syntax.EncoderOps
import io.circe.{Json, Decoder, Encoder, HCursor, ParsingFailure}
import io.circe.generic.extras.{Configuration, ConfiguredJsonCodec, JsonKey}
import java.util.jar.Manifest
import main.MainClass
import java.util.Scanner

class Harness {
  val NOT_IMPLEMENTED: String = "This case is not yet implemented."

  val SKIPPED_CASES: Map[String, String] = Map(
    "escaped pointer ref" -> NOT_IMPLEMENTED,
    "empty tokens in $ref json-pointer" -> NOT_IMPLEMENTED,
    "schema that uses custom metaschema with with no validation vocabulary" -> NOT_IMPLEMENTED,
    "small multiple of large integer" -> NOT_IMPLEMENTED,
    "$ref to $ref finds detached $anchor" -> NOT_IMPLEMENTED,
    "$ref to $dynamicRef finds detached $dynamicAnchor" -> NOT_IMPLEMENTED
  )

  val SKIPPED_TESTS: Map[String, TestSkip] = Map(
    "minLength validation" -> TestSkip("one supplementary Unicode code point is not long enough", NOT_IMPLEMENTED),
    "maxLength validation" -> TestSkip("two supplementary Unicode code points is long enough", NOT_IMPLEMENTED)
  )

  implicit val customConfig: Configuration = Configuration.default.withDefaults

  def operate(line: String) = {
    try {
      val node: Json = parse(line) match {
        case Right(json)          => json
        case Left(parsingFailure) => throw parsingFailure
      }

      val cmd: String = (node \\ "cmd").headOption match {
        case Some(value) => value.asString.getOrElse("")
        case None        => throw new RuntimeException("Failed to get cmd")
      }
      cmd match {
        case "start"   => println(start(node))
        case "dialect" => println(dialect(node))
        case "run"     => println(run(node))
        case "stop"    => System.exit(0)
        case default   => throw new IllegalArgumentException(s"Unknown command: $default")
      }
    } catch {
      case e: Exception =>
        // TODO: Check if this is the correct way to handle errors
        println(Errored(e.getMessage(), e.getStackTrace().mkString("\n")).asJson.noSpaces)
    }
  }

  def start(node: Json): String = {
    val startRequest: StartRequest = decodeTo[StartRequest](node)
    val version = startRequest.version
    if (version == 1) {
      Harness.started = true
      val is: InputStream = getClass.getResourceAsStream("META-INF/MANIFEST.MF")
      val attributes = new Manifest(is).getMainAttributes

      val dialects: Json = Json.arr(Json.fromString("https://json-schema.org/draft/2020-12/schema"))

      val implementation: Json = Json.obj(
        "language" -> Json.fromString("scala"),
        "name" -> Json.fromString(attributes.getValue("Implementation-Name")),
        "version" -> Json.fromString(attributes.getValue("Implementation-Version")),
        "homepage" -> Json.fromString("https://gitlab.lip6.fr/jsonschema/modernjsonschemavalidator"),
        "issues" -> Json.fromString("https://gitlab.lip6.fr/jsonschema/modernjsonschemavalidator/issues"),
        "dialects" -> dialects,
        "os" -> Json.fromString(System.getProperty("os.name")),
        "os_version" -> Json.fromString(System.getProperty("os.version")),
        "language_version" -> Json.fromString(Runtime.version().toString)
      )

      StartResponse(version, true, implementation).asJson.noSpaces
    } else {
      // TODO: should return an errored response?
      throw new IllegalArgumentException(s"Unsupported version: ${startRequest.version}")
    }
  }

  def dialect(node: Json): String = {
    if (!Harness.started) {
      throw new RuntimeException("Bowtie hasn't started!")
    }
    val dialectRequest: DialectRequest = decodeTo[DialectRequest](node)
    // TODO: Handle properly (if this is not correct); We should probably return true
    DialectResponse(false).asJson.noSpaces
  }

  def run(node: Json): String = {
    if (!Harness.started) {
      throw new RuntimeException("Bowtie hasn't started!")
    }

    val runRequest: RunRequest = decodeTo[RunRequest](node)
    try {
      val caseDescription = runRequest.testCase.description
      if (SKIPPED_CASES.contains(caseDescription)) {
        return SkippedRunResponse(runRequest.seq, true, Some(NOT_IMPLEMENTED)).asJson.noSpaces
      }

      val registryMap: Map[String, String] = runRequest.testCase.registry
        .flatMap { node =>
          node.as[Map[String, Json]].toOption.map { jsonMap => jsonMap.mapValues(_.noSpaces).toMap }
          node.as[Map[String, Json]].toOption.map { jsonMap => jsonMap.mapValues(_.noSpaces).toMap }
        }
        .getOrElse(null)

      var resultArray = Vector.empty[Json]
      runRequest.testCase.tests.foreach { test =>
        try {
          val testDescription: String = test.description
          val instance: String = test.instance.noSpaces

          if (SKIPPED_TESTS.contains(caseDescription) && SKIPPED_TESTS(caseDescription).description == testDescription) {
            resultArray :+= SkippedTest(message = Some(SKIPPED_TESTS(caseDescription).message)).asJson
          } else {
            val schema: String = runRequest.testCase.schema.noSpaces

            val result: Json = Json.obj("valid" -> MainClass.validateInstance(schema, instance, registryMap).asJson)
            resultArray :+= result
          }
        } catch {
          // TODO: Testing
          case e: Exception => {
            val error: Errored = Errored(e.getMessage(), e.getStackTrace().mkString("\n"))
            resultArray :+= ErroredTest(true, error).asJson
          }
        }
      }
      RunResponse(runRequest.seq, resultArray).asJson.noSpaces

    } catch {
      // TODO: We currently abort instead of creating errored responses?
      case e: Exception =>
        val error: Errored = Errored(e.getMessage(), e.getStackTrace().mkString("\n"))
        ErroredRunResponse(runRequest.seq, context = error).asJson.noSpaces
    }
  }

  def decodeTo[Request: Decoder](json: Json): Request = {
    json.as[Request] match {
      case Right(value)          => value
      case Left(decodingFailure) => throw decodingFailure
    }
  }
}

// TODO: Check if we need any more case classes

case class Test(description: String, comment: Option[String], instance: Json, valid: Option[Boolean])

case class TestCase(description: String, comment: Option[String], schema: Json, registry: Option[Json], tests: List[Test])

case class StartRequest(version: Int)

case class StartResponse(version: Int, ready: Boolean, implementation: Json)

case class DialectRequest(dialect: String)

case class DialectResponse(ok: Boolean)

@ConfiguredJsonCodec
case class RunRequest(seq: Json, @JsonKey("case") testCase: TestCase)

object RunRequest {
  implicit val config: Configuration = Configuration.default.withSnakeCaseMemberNames.withDefaults
}

case class RunResponse(seq: Json, results: Vector[Json])

case class ErroredRunResponse(seq: Json, errored: Boolean = true, context: Errored)

case class ErroredTest(errored: Boolean = true, context: Errored) // TODO: restructure?

case class Errored(message: String, traceback: String) // TODO: needed?

case class SkippedRunResponse(seq: Json, skipped: Boolean = true, message: Option[String] = None)

case class SkippedTest(skipped: Boolean = true, message: Option[String] = None)

case class TestSkip(description: String, message: String)

object Harness {
  var started: Boolean = false

  def main(args: Array[String]): Unit = {
    val input = new Scanner(System.in)
    while (true) {
      new Harness().operate(input.nextLine())
    }
  }
}
